<img src="https://github.com/PythonFreeCourse/Notebooks/blob/main/content/week05/images/logo.jpg?raw=1" style="display: block; margin-left: auto; margin-right: auto;" alt="לוגו של מיזם לימוד הפייתון. נחש מצויר בצבעי צהוב וכחול, הנע בין האותיות של שם הקורס: לומדים פייתון. הסלוגן המופיע מעל לשם הקורס הוא מיזם חינמי ללימוד תכנות בעברית.">

# <span style="text-align: right; direction: rtl; float: right;">Generators</span>

## <span style="text-align: right; direction: rtl; float: right; clear: both;">הקדמה</span>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    הפונקציות שיצרנו עד כה נבנו כך שיחזירו ערך אחד בכל קריאה.<br>
    הערך הזה יכול היה להיות מכל טיפוס שהוא: בוליאני, מחרוזת, tuple וכדומה.<br>
    אם נרצה להחזיר כמה ערכים יחד, תמיד נוכל להחזיר אותם כרשימה או כ־tuple.<br>
    אבל מה קורה כשאנחנו רוצים להחזיר סדרת ערכים גדולה מאוד או אפילו אין־סופית?
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    למשל:
</p>

<ul style="text-align: right; direction: rtl; float: right; clear: both;">
    <li>הכתובות של כל הדפים הקיימים באינטרנט.</li>
    <li>מילות כל השירים שראו אור מאז שנת 1400 לספירה.</li>
    <li>כל המספרים השלמים הגדולים מ־0.</li>
</ul>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    בכלים שיש בידינו כרגע, נמצא שיש בעיה ליצור רשימות כאלו.<br>
    עבור רשימות גדולות מאוד – לא יהיה למחשב די זיכרון ולבסוף הוא ייכשל בשמירת ערכים חדשים.<br>
    ומה בנוגע לרשימות אין־סופיות? הן... ובכן... אין־סופיות, ולכן מלכתחילה לא נוכל ליצור אותן.
</p>

## <span style="text-align: right; direction: rtl; float: right; clear: both;">הגדרה</span>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    פתרון שמעניין לחשוב עליו הוא "פונקציה עצלנית".<br>
    אם אנחנו בשום שלב לא יכולים להחזיק בזיכרון המחשב את כל האיברים (כי יש יותר מדי מהם, או כי זו סדרה אין־סופית),<br>
    אולי נוכל לשלוח תחילה את הערך הראשון – ואת הערכים שבאים אחריו נשלח רק כשיבקשו אותם מאיתנו.
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    פונקציה עצלנית שכזו נקראת <dfn>generator</dfn>, ובכל פעם שנבקש ממנה, היא תחזיר לנו איבר יחיד מתוך סדרת ערכים.<br>
    תחילה – היא תחזיר רק את הערך הראשון, בלי לחשב את שאר האיברים. אחר כך, באותו אופן, רק את השני, אחר כך רק את השלישי וכן הלאה.<br>
    ההבדל העיקרי בין generator לבין פונקציה רגילה, הוא שב־generator נבחר להחזיר את הערכים אחד־אחד ולא תחת מבנה מאוגד.
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    נסכם: generator היא פונקציה שיוצרת עבורנו בכל פעם ערך אחד, מחזירה אותו, ומחכה עד שנבקש את האיבר הבא.
</p>

## <span style="text-align: right; direction: rtl; float: right; clear: both;">שימוש</span>

### <span style="text-align: right; direction: rtl; float: right; clear: both;">יצירת generator בסיסי</span>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    נתחיל בהגדרת generator מטופש למדי:
</p>

In [1]:
def silly_generator():
    a = 1
    yield a
    b = a + 1
    yield b
    c = [1, 2, 3]
    yield c

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    מעניין! זה נראה ממש כמו פונקציה. נקרא למבנה הזה שיצרנו "<dfn>פונקציית ה־generator</dfn>".<br>
    אבל מהו ה־<code>yield</code> המוזר הזה שנמצא שם?
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    לפני שנתהה על קנקנו, בואו ננסה לקרוא לפונקציה ונראה מה היא מחזירה:
</p>

In [2]:
print(silly_generator())

<generator object silly_generator at 0x7d24780187b0>


<p style="text-align: right; direction: rtl; float: right; clear: both;">
    אומנם למדנו שלא אומרים איכס על פונקציות, אבל מה קורה פה?
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    בניגוד לפונקציות רגילות, קריאה ל־generator לא מחזירה ערך מייד.<br>
    במקום ערך היא מחזירה מעין סמן, כמו בקובץ, שאפשר לדמיין כחץ שמצביע על השורה הראשונה של הפונקציה.<br>
    נשמור את הסמן על משתנה:
</p>

In [3]:
our_generator = silly_generator()

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    בעקבות הקריאה ל־<code>silly_generator</code> נוצר לנו סמן שמצביע כרגע על השורה <code>a = 1</code>.<br>
    המינוח המקצועי לסמן הזה הוא <dfn>generator iterator</dfn>.
</p>

<img src="https://github.com/PythonFreeCourse/Notebooks/blob/main/content/week05/images/silly_generator1.png?v=1&raw=1" width="300px" style="display: block; margin-left: auto; margin-right: auto;" alt="תוכן הפונקציה silly_generator, כאשר חץ מצביע לשורה הראשונה שלה – a = 1">

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    אחרי שהרצנו את השורה <code dir="ltr">our_generator = silly_generator()</code>, הסמן המדובר נשמר במשתנה בשם <var>our_generator</var>.<br>
    זה זמן מצוין לבקש מה־generator להחזיר ערך.<br>
    נעשה זאת בעזרת הפונקציה הפייתונית <code>next</code>:
</p>

In [4]:
next_value = next(our_generator)
print(next_value)

1


<p style="text-align: right; direction: rtl; float: right; clear: both;">
    כדי להבין מה התרחש נצטרך להבין שני דברים חשובים שקשורים ל־generators:<br>
</p>
<ol style="text-align: right; direction: rtl; float: right; clear: both;">
    <li>קריאה ל־<code>next</code> היא כמו לחיצה על "נגן" (Play) – היא גורמת לסמן לרוץ עד שהוא מגיע לשורה של החזרת ערך.</li>
    <li>מילת המפתח <code>yield</code> דומה למילת המפתח <code>return</code> – היא מפסיקה את ריצת הסמן, ומחזירה את הערך שמופיע אחריה.</li>
</ol>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    אז היה לנו סמן שהצביע על השורה הראשונה. לחצנו Play, והוא הריץ את הקוד עד שהוא הגיע לנקודה שבה מחזירים ערך.<br>
    ההבדל בין פונקציה לבין generator, הוא ש<mark>כשאנחנו מחזירים ערך בעזרת <code>yield</code> אנחנו "מקפיאים" את המצב שבו
    יצאנו מהפונקציה.</mark><br>
    ממש כמו ללחוץ על "Pause".<br>
    כשנקרא ל־<code>next</code> בפעם הבאה – הפונקציה תמשיך לרוץ מאותו המקום שבו השארנו את הסמן, עם אותם ערכי משתנים.<br>
    עכשיו הסמן מצביע על השורה <code>b = a + 1</code>, ומחכה שמישהו יקרא שוב ל־<code>next</code> כדי שהפונקציה תוכל להמשיך לרוץ:
</p>

In [5]:
print(next(our_generator))

2


<p style="text-align: right; direction: rtl; float: right; clear: both;">
    נסכם מה קרה עד עכשיו:
</p>

<ol style="text-align: right; direction: rtl; float: right; clear: both;">
    <li>הגדרנו פונקציה בשם <var>silly_generator</var>, שאמורה להחזיר את הערכים <samp>1</samp>, <samp>2</samp> ו־<samp dir="ltr">[1, 2, 3]</samp>. קראנו לה "<em>פונקציית הגנרטור</em>".</li>
    <li>בעזרת קריאה לפונקציית הגנרטור, יצרנו "סמן" (generator iterator) שנקרא <var>our_generator</var> ומצביע לשורה הראשונה בפונקציה.</li>
    <li>בעזרת קריאה ל־<code>next</code> על ה־generator iterator, הרצנו את הסמן עד שה־generator החזיר ערך.</li>
    <li>למדנו ש־generator־ים מחזירים ערכים בעיקר בעזרת yield – שמחזיר ערך ושומר את המצב שבו הפונקציה עצרה.</li>
    <li>קראנו שוב ל־<code>next</code> על ה־generator iterator, וראינו שהוא ממשיך מהמקום שבו ה־generator הפסיק לרוץ פעם קודמת.</li>
</ol>

<div class="align-center" style="display: flex; text-align: right; direction: rtl; clear: both;">
    <div style="display: flex; width: 10%; float: right; clear: both;">
        <img src="https://github.com/PythonFreeCourse/Notebooks/blob/main/content/week05/images/exercise.svg?raw=1" style="height: 50px !important;" alt="תרגול">
    </div>
    <div style="width: 70%">
        <p style="text-align: right; direction: rtl; float: right; clear: both;">
            תוכלו לחזות מה יקרה אם נקרא שוב ל־<code>next(our_generator)</code>?
        </p>
    </div>
    <div style="display: flex; width: 20%; border-right: 0.1rem solid #A5A5A5; padding: 1rem 2rem;">
        <p style="text-align: center; direction: rtl; justify-content: center; align-items: center; clear: both;">
            <strong>חשוב!</strong><br>
            פתרו לפני שתמשיכו!
        </p>
    </div>
</div>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    ננסה:
</p>

In [6]:
print(next(our_generator))

[1, 2, 3]


<p style="text-align: right; direction: rtl; float: right; clear: both;">
    יופי! הכל הלך כמצופה.<br>
    אבל מה צופן לנו העתיד?<br>
    בפעם הבאה שנבקש ערך מהפונקציה, הסמן שלנו ירוץ הלאה ולא ייתקל ב־<code>yield</code>.<br>
    במקרה כזה, נקבל שגיאת <var>StopIteration</var>, שמבשרת לנו ש־<code>next</code> לא הצליח לחלץ מה־generator את הערך הבא.
</p>

In [7]:
print(next(our_generator))

StopIteration: 

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    מובן שאין סיבה להילחץ.<br>
    במקרה הזה אפילו לא מדובר במשהו רע – פשוט כילינו את כל הערכים מה־generator iterator שלנו.<br>
    פונקציית ה־generator עדיין קיימת!<br>
    אפשר ליצור עוד generator iterator אם נרצה, ולקבל את כל הערכים שנמצאים בו באותה צורה:
</p>

In [8]:
our_generator = silly_generator()
print(next(our_generator))
print(next(our_generator))
print(next(our_generator))

1
2
[1, 2, 3]


<p style="text-align: right; direction: rtl; float: right; clear: both;">
    אבל כשחושבים על זה, זה קצת מגוחך.<br>
    בכל פעם שנרצה להשיג את הערך הבא נצטרך לרשום <code>next</code>?<br>
    חייבת להיות דרך טובה יותר!
</p>

### <span style="text-align: right; direction: rtl; float: right; clear: both;">כל generator הוא גם iterable</span>

#### <span style="text-align: right; direction: rtl; float: right; clear: both;">for</span>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    אז למעשה, יש יותר מדרך טובה אחת להשיג את כל הערכים שיוצאים מ־generator מסוים.<br>
    כהקדמה, נניח פה עובדה שלא תשאיר אתכם אדישים: ה־generator iterator הוא... iterable! הפתעת השנה, אני יודע!<br>
    אמנם אי אפשר לפנות לאיברים שלו לפי מיקום, אך בהחלט אפשר לעבור עליהם בעזרת לולאת <code>for</code>, לדוגמה:
</p>

In [9]:
our_generator = silly_generator()
for item in our_generator:
    print(item)

1
2
[1, 2, 3]


<p style="text-align: right; direction: rtl; float: right; clear: both;">
    מה מתרחש כאן?<br>
    אנחנו מבקשים מלולאת ה־<code>for</code> לעבור על ה־generator iterator שלנו.<br>
    ה־<code>for</code> עושה עבורנו את העבודה אוטומטית:
</p>

<ol style="text-align: right; direction: rtl; float: right; clear: both;">
    <li>הוא מבקש את האיבר הבא מה־generator iterator באמצעות <code>next</code>.</li>
    <li>הוא מכניס את האיבר שהוא קיבל מה־generator ל־<var>item</var>.</li>
    <li>הוא מבצע את גוף הלולאה פעם אחת עבור האיבר שנמצא ב־<var>item</var>.</li>
    <li>הוא חוזר לראש הלולאה שוב, ומנסה לקבל את האיבר הבא באמצעות <code>next</code>. כך עד שייגמרו האיברים ב־generator iterator.</li>
</ol>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    שימו לב שהעובדות שלמדנו בנוגע לאותו "סמן" יבואו לידי ביטוי גם כאן.<br>
    הרצה נוספת של הלולאה על אותו סמן לא תדפיס יותר איברים, כיוון שהסמן מצביע כעת על סוף פונקציית ה־generator:
</p>

In [10]:
for item in our_generator:
    print(item)

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    למזלנו, לולאות <code>for</code> יודעות לטפל בעצמן בשגיאת <code>StopIteration</code>, ולכן שגיאה שכזו לא תקפוץ לנו במקרה הזה.
</p>

#### <span style="text-align: right; direction: rtl; float: right; clear: both;">המרת טיפוסים</span>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    דרך אחרת, לדוגמה, היא לבקש להמיר את ה־generator iterator לסוג משתנה אחר שהוא גם iterable:
</p>

In [11]:
our_generator = silly_generator()
items = list(our_generator)
print(items)

[1, 2, [1, 2, 3]]


<p style="text-align: right; direction: rtl; float: right; clear: both;">
    בקוד שלמעלה, השתמשנו בפונקציה <code>list</code> שיודעת להמיר ערכים iterable־ים לרשימות.<br>
    שימו לב שמה שלמדנו בנוגע ל"סמן" יבוא לידי ביטוי גם בהמרות:
</p>

In [12]:
print(list(our_generator))

[]


### <span style="text-align: right; direction: rtl; float: right; clear: both;">שימושים פרקטיים</span>

#### <span style="text-align: right; direction: rtl; float: right; clear: both;">חיסכון בזיכרון</span>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
נכתוב פונקציה רגילה שמקבלת מספר שלם, ומחזירה רשימה של כל המספרים השלמים מ־0 ועד אותו מספר (נשמע לכם מוכר?):
</p>

In [ ]:
def my_range(upper_limit):
    numbers = []
    current_number = 0
    while current_number < upper_limit:
        numbers.append(current_number)
        current_number = current_number + 1
    return numbers


for number in my_range(1000):
    print(number)

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    בפונקציה הזו אנחנו יוצרים רשימת מספרים חדשה, המכילה את כל המספרים בין 0 לבין המספר שהועבר לפרמטר <var>upper_limit</var>.<br>
    אך ישנה בעיה חמורה – הפעלת הפונקציה גורמת לניצול משאבים רבים!<br>
    אם נכניס כארגומנט 1,000 – נצטרך להחזיק רשימה המכילה 1,000 איברים שונים, ואם נכניס מספר גדול מדי – עלול להיגמר לנו הזיכרון.
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    אבל איזו סיבה יש לנו להחזיק בזיכרון את רשימת כל המספרים?<br>
    אם לא עולה צורך מובהק שכזה, ייתכן שעדיף להחזיק בזיכרון מספר אחד בלבד בכל פעם, ולהחזירו מייד בעזרת generator:
</p>

In [13]:
def my_range(upper_limit):
    current_number = 0
    while current_number < upper_limit:
        yield current_number
        current_number = current_number + 1


our_generator = my_range(1000)
for number in our_generator:
    print(number)

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249
250
251
252
253
254
255
256
257
258
259
260
261
262
263
264
265
266
267
268
269
270
271
272
273
274
275
276
27

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    שימו לב כמה הגרסה הזו אלגנטית יותר!<br>
    בכל פעם אנחנו פשוט שולחים את ערכו של מספר אחד (<var>current_number</var>) החוצה.<br>
    כשמבקשים את הערך הבא מה־generator iterator, פונקציית ה־generator חוזרת לעבוד מהנקודה שבה היא עצרה:<br>
    היא מעלה את ערכו של המספר הנוכחי, בודקת אם הוא נמוך מ־<var>upper_limit</var>, ושולחת גם אותו החוצה.<br>
    בשיטה הזו, <code>my_range(numbers)</code> לא מחזירה לנו רשימה של התוצאות – אלא generator iterator שמחזיר ערך אחד בכל פעם.<br>
    כך אנחנו לעולם לא מחזיקים בזיכרון 1,000 מספרים בו־זמנית.
</p>

<div class="align-center" style="display: flex; text-align: right; direction: rtl; clear: both;">
    <div style="display: flex; width: 10%; float: right; clear: both;">
        <img src="https://github.com/PythonFreeCourse/Notebooks/blob/main/content/week05/images/exercise.svg?raw=1" style="height: 50px !important;" alt="תרגול">
    </div>
    <div style="width: 70%">
        <p style="text-align: right; direction: rtl; float: right; clear: both;">
            לפניכם פונקציה שמקבלת רשימה, ומחזירה עבור כל מספר ברשימה את ערכו בריבוע.<br>
            זוהי גרסה מעט בזבזנית שמשתמשת בהרבה זיכרון. תוכלו להמיר אותה להיות generator?
        </p>
    </div>
    <div style="display: flex; width: 20%; border-right: 0.1rem solid #A5A5A5; padding: 1rem 2rem;">
        <p style="text-align: center; direction: rtl; justify-content: center; align-items: center; clear: both;">
            <strong>חשוב!</strong><br>
            פתרו לפני שתמשיכו!
        </p>
    </div>
</div>

In [ ]:
def square_numbers(numbers):
    squared_numbers = []
    for number in numbers:
        squared_numbers.append(number ** 2)
    return squared_numbers


for number in square_numbers(my_range(1000)):
    print(number)

In [18]:
def square_numbers(numbers):
  for number in numbers:
     yield number ** 2



our_generator = square_numbers(my_range(1000))
for number in our_generator:
    print(number)

0
1
4
9
16
25
36
49
64
81
100
121
144
169
196
225
256
289
324
361
400
441
484
529
576
625
676
729
784
841
900
961
1024
1089
1156
1225
1296
1369
1444
1521
1600
1681
1764
1849
1936
2025
2116
2209
2304
2401
2500
2601
2704
2809
2916
3025
3136
3249
3364
3481
3600
3721
3844
3969
4096
4225
4356
4489
4624
4761
4900
5041
5184
5329
5476
5625
5776
5929
6084
6241
6400
6561
6724
6889
7056
7225
7396
7569
7744
7921
8100
8281
8464
8649
8836
9025
9216
9409
9604
9801
10000
10201
10404
10609
10816
11025
11236
11449
11664
11881
12100
12321
12544
12769
12996
13225
13456
13689
13924
14161
14400
14641
14884
15129
15376
15625
15876
16129
16384
16641
16900
17161
17424
17689
17956
18225
18496
18769
19044
19321
19600
19881
20164
20449
20736
21025
21316
21609
21904
22201
22500
22801
23104
23409
23716
24025
24336
24649
24964
25281
25600
25921
26244
26569
26896
27225
27556
27889
28224
28561
28900
29241
29584
29929
30276
30625
30976
31329
31684
32041
32400
32761
33124
33489
33856
34225
34596
34969
35344
35721
36100


#### <span style="text-align: right; direction: rtl; float: right; clear: both;">תשובות חלקיות</span>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    לעיתים ניאלץ לבצע חישוב ארוך, שהשלמתו תימשך זמן רב מאוד.<br>
    במקרה כזה, נוכל להשתמש ב־generator כדי לקבל חלק מהתוצאה בזמן אמת,<br>
    בזמן שבפונקציה "רגילה" נצטרך להמתין עד סיום החישוב כולו.
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    שלשה פיתגורית, לדוגמה, היא שלישיית מספרים שלמים וחיוביים, $a$, $b$ ו־$c$, שעונים על הדרישה $a^2 + b^2 = c^2$.<br>
    אם כך, כדי ששלושה מספרים שאנחנו בוחרים ייחשבו שלשה פיתגורית,<br>
    הסכום של ריבוע המספר הראשון וריבוע המספר השני, אמור להיות שווה לערכו של המספר השלישי בריבוע.
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    אלו דוגמאות לשלשות פיתגוריות:
</p>

<ul style="text-align: right; direction: rtl; float: right; clear: both;">
    <li>$(3, 4, 5)$, כיוון ש־$9 + 16 = 25$.<br>
        9 הוא 3 בריבוע, 16 הוא 4 בריבוע ו־25 הוא 5 בריבוע.
    </li>
    <li>$(5, 12, 13)$, כיוון ש־$25 + 144 = 169$.</li>
    <li>$(8, 15, 17)$, כיוון ש־$64 + 225 = 289$.</li>
</ul>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    ננסה למצוא את כל השלשות הפיתגוריות מתחת ל־10,000 בעזרת קוד שרץ על כל השלשות האפשריות:
</p>

In [19]:
def find_pythagorean_triples(upper_bound=10_000):
    pythagorean_triples = []
    for c in range(3, upper_bound):
        for b in range(2, c):
            for a in range(1, b):
                if a ** 2 + b **2 == c ** 2:
                    pythagorean_triples.append((a, b, c))
    return pythagorean_triples


for triple in find_pythagorean_triples():
    print(triple)

KeyboardInterrupt: 

<div class="align-center" style="display: flex; text-align: right; direction: rtl;">
    <div style="display: flex; width: 10%; float: right; ">
        <img src="https://github.com/PythonFreeCourse/Notebooks/blob/main/content/week05/images/warning.png?raw=1" style="height: 50px !important;" alt="אזהרה!">
    </div>
    <div style="width: 90%">
        <p style="text-align: right; direction: rtl;">
            הרצת התא הקודם תתקע את המחברת (חישוב התוצאה יימשך זמן רב).<br>
            כדי להיות מסוגלים להריץ את התאים הבאים, לחצו <samp>00</samp> לאחר הרצת התא, ובחרו <em>Restart</em>.<br>
            אל דאגה – האתחול יתבצע אך ורק עבור המחברת, ולא עבור מחשב.
        </p>
    </div>
</div>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    יו, כמה זמן נמשכת הרצת הקוד הזה... 😴<br>
    הלוואי שעד שהקוד הזה היה מסיים היינו מקבלים לפחות <em>חלק</em> מהתוצאות!<br>
    נפנה ל־generator־ים לעזרה:
</p>

In [20]:
def find_pythagorean_triples(upper_bound=10_000):
    for c in range(3, upper_bound):
        for b in range(2, c):
            for a in range(1, b):
                if a ** 2 + b **2 == c ** 2:
                    yield a, b, c


for triple in find_pythagorean_triples():
    print(triple)

(3, 4, 5)
(6, 8, 10)
(5, 12, 13)
(9, 12, 15)
(8, 15, 17)
(12, 16, 20)
(15, 20, 25)
(7, 24, 25)
(10, 24, 26)
(20, 21, 29)
(18, 24, 30)
(16, 30, 34)
(21, 28, 35)
(12, 35, 37)
(15, 36, 39)
(24, 32, 40)
(9, 40, 41)
(27, 36, 45)
(30, 40, 50)
(14, 48, 50)
(24, 45, 51)
(20, 48, 52)
(28, 45, 53)
(33, 44, 55)
(40, 42, 58)
(36, 48, 60)
(11, 60, 61)
(39, 52, 65)
(33, 56, 65)
(25, 60, 65)
(16, 63, 65)
(32, 60, 68)
(42, 56, 70)
(48, 55, 73)
(24, 70, 74)
(45, 60, 75)
(21, 72, 75)
(30, 72, 78)
(48, 64, 80)
(18, 80, 82)
(51, 68, 85)
(40, 75, 85)
(36, 77, 85)
(13, 84, 85)
(60, 63, 87)
(39, 80, 89)
(54, 72, 90)
(35, 84, 91)
(57, 76, 95)
(65, 72, 97)
(60, 80, 100)
(28, 96, 100)
(20, 99, 101)
(48, 90, 102)
(40, 96, 104)
(63, 84, 105)
(56, 90, 106)
(60, 91, 109)
(66, 88, 110)
(36, 105, 111)
(15, 112, 113)
(69, 92, 115)
(80, 84, 116)
(45, 108, 117)
(56, 105, 119)
(72, 96, 120)
(22, 120, 122)
(27, 120, 123)
(75, 100, 125)
(44, 117, 125)
(35, 120, 125)
(78, 104, 130)
(66, 112, 130)
(50, 120, 130)
(32, 126, 13

KeyboardInterrupt: 

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    איך זה קרה? קיבלנו את התשובה בתוך שבריר שנייה!<br>
    ובכן, זה לא מדויק – קיבלנו חלק מהתשובות. שימו לב שהקוד ממשיך להדפיס :)<br>
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    להזכירכם, ה־generator שולח את התוצאה החוצה מייד כשהוא מוצא שלשה אחת,<br>
    וה־for מקבל מה־generator iterable כל שלשה ברגע שהיא נמצאה.<br>
    ברגע שה־for מקבל שלשה, הוא מבצע את גוף הלולאה עבור אותה שלשה, ורק אז מבקש מ־generator את האיבר הבא.<br>
    בגלל האופי של generators, הקוד בתא האחרון מדפיס לנו כל שלשה ברגע שהוא מצא אותה, ולא מחכה עד שיימצאו כל השלשות.
</p>

#### <span style="text-align: right; direction: rtl; float: right; clear: both;">תרגול ביניים: מספרים פראיים</span>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    "פירוק לגורמים של מספר שלם" היא בעיה שחישוב פתרונה נמשך זמן רב במחשבים מודרניים.
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    עליכם לכתוב פונקציה שמקבלת מספר חיובי שלם $n$, ומחזירה קבוצת מספרים שמכפלתם (תוצאת הכפל ביניהם) היא $n$.<br>
    לדוגמה, המספר 1,386 בנוי מהמכפלה של קבוצת המספרים $2 \cdot 3 \cdot 3 \cdot 7 \cdot 11$.<br>
    כל מספר בקבוצת המספרים הזו חייב להיות ראשוני.<br>
    להזכירכם: מספר ראשוני הוא מספר שאין לו מחלקים חוץ מעצמו ומ־1.
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    הניחו שהמספר שהתקבל אינו ראשוני.<br>
    מה היתרון של generator על פני פונקציה רגילה שעושה אותו דבר?
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">    
    רמז: <span style="background: black; color: black;">אם תנסו לחלק את המספר ב־2, ואז ב־3 (וכן הלאה), בסופו של דבר תגיעו למחלק ראשוני של המספר.</span><br>
    רמז עבה: <span style="background: black; color: black;">בכל פעם שמצאתם מחלק אחד למספר, חלקו את המספר במחלק, והתחילו את החיפוש מחדש. מתי עליכם לעצור?</span>
</p>

In [43]:
def decompose_to_primary(num):
  while num > 1:
    i = 2
    found_primary = False
    while not found_primary:
      if (num / i).is_integer():
        num /= i
        found_primary = True
        yield i

      i += 1



In [44]:
for prime in decompose_to_primary(1386):
    print(prime)


2
3
3
7
11


#### <span style="text-align: right; direction: rtl; float: right; clear: both;">אוספים אין־סופיים</span>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    עבור בעיות מסוימות, נרצה להיות מסוגלים להחזיר אין־סוף תוצאות.<br>
    ניקח כדוגמה לסדרה אין־סופית את סדרת פיבונאצ'י, שבה כל איבר הוא סכום זוג האיברים הקודמים לו:<br>
    $1, 1, 2, 3, 5, 8, \ldots$
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    נממש פונקציה שמחזירה לנו את סדרת פיבונאצ'י.<br>
    בפונקציה רגילה אין לנו אפשרות להחזיר מספר אין־סופי של איברים, ולכן נצטרך להחליט על מספר האיברים המרבי שנרצה להחזיר:
</p>

In [ ]:
def fibonacci(max_items):
    a = 1
    b = 1
    numbers = [1, 1]
    while len(numbers) < max_items:
        a, b = b, a + b  # Unpacking
        numbers.append(b)
    return numbers


for number in fibonacci(10):
    print(number)

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    לעומת זאת, ל־generators לא חייב להיות סוף מוגדר.<br>
    נשתמש ב־<code>while True</code> שתמיד מתקיים, כדי שבסופו של דבר – תמיד נגיע ל־<code>yield</code>:
</p>

In [45]:
def fibonacci():
    a = 1
    b = 1
    numbers = [1, 1]
    while True:  # תמיד מתקיים
        yield a
        a, b = b, a + b


generator_iterator = fibonacci()
for number in range(10):
    print(next(generator_iterator))

# אני יכול לבקש בקלות רבה את 10 האיברים הבאים בסדרה
for number in range(10):
    print(next(generator_iterator))

1
1
2
3
5
8
13
21
34
55
89
144
233
377
610
987
1597
2584
4181
6765


<div class="align-center" style="display: flex; text-align: right; direction: rtl;">
    <div style="display: flex; width: 10%; float: right; ">
        <img src="https://github.com/PythonFreeCourse/Notebooks/blob/main/content/week05/images/warning.png?raw=1" style="height: 50px !important;" alt="אזהרה!">
    </div>
    <div style="width: 90%">
        <p style="text-align: right; direction: rtl;">
            generators אין־סופיים יכולים לגרום בקלות ללולאות אין־סופיות, גם בלולאות <code>for</code>.<br>
            שימו לב לצורת ההתעסקות העדינה בדוגמאות למעלה.<br>
            הרצת לולאת <code>for</code> ישירות על ה־generator iterator הייתה מכניסה אותנו ללולאה אין־סופית.
        </p>
    </div>
</div>

<div class="align-center" style="display: flex; text-align: right; direction: rtl; clear: both;">
    <div style="display: flex; width: 10%; float: right; clear: both;">
        <img src="https://github.com/PythonFreeCourse/Notebooks/blob/main/content/week05/images/exercise.svg?raw=1" style="height: 50px !important;" alt="תרגול">
    </div>
    <div style="width: 70%">
        <p style="text-align: right; direction: rtl; float: right; clear: both;">
            כתבו generator שמחזיר את כל המספרים השלמים הגדולים מ־0.
        </p>
    </div>
    <div style="display: flex; width: 20%; border-right: 0.1rem solid #A5A5A5; padding: 1rem 2rem;">
        <p style="text-align: center; direction: rtl; justify-content: center; align-items: center; clear: both;">
            <strong>חשוב!</strong><br>
            פתרו לפני שתמשיכו!
        </p>
    </div>
</div>

In [52]:
def naturals_generator():
  i = 0
  while True:
    i += 1
    yield i


In [53]:
g = naturals_generator()
for _ in range(10):
  print(next(g))

1
2
3
4
5
6
7
8
9
10


### <span style="text-align: right; direction: rtl; float: right; clear: both;">ריבוי generator iterators</span>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    נגדיר generator פשוט שמחזיר את האיברים <samp>1</samp>, <samp>2</samp> ו־<samp>3</samp>:
</p>

In [54]:
def simple_generator():
    yield 1
    yield 2
    yield 3

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    ניצור שני generator iterators ("סמנים") שונים שמצביעים לשורה הראשונה של ה־generator שמופיע למעלה:
</p>

In [55]:
first_gen = simple_generator()
second_gen = simple_generator()

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    בעניין זה, חשוב להבין שכל אחד מה־generator iterators הוא "חץ" נפרד שמצביע לשורה הראשונה ב־<var>simple_generator</var>.<br>
    אם נבקש מכל אחד מהם להחזיר ערך, נקבל משניהם את 1, ואותו חץ דמיוני יעבור בשני ה־generator iterators להמתין בשורה השנייה:
</p>

In [56]:
print(next(first_gen))
print(next(second_gen))

1
1


<p style="text-align: right; direction: rtl; float: right; clear: both;">
    נוכל לקדם את <var>first_gen</var>, לדוגמה, לסוף הפונקציה:
</p>

In [57]:
print(next(first_gen))
print(next(first_gen))

2
3


<p style="text-align: right; direction: rtl; float: right; clear: both;">
    אבל <var>second_gen</var> הוא חץ נפרד, שעדיין מצביע לשורה השנייה של פונקציית ה־generator.<br>
    אם נבקש ממנו את הערך הבא, הוא ימשיך את המסע מהערך <samp>2</samp>:<br>
</p>

In [58]:
print(next(second_gen))

2


<p style="text-align: right; direction: rtl; float: right; clear: both;">
    ממצב זה נוכל להסיק שאפשר ליצור יותר מ־generator iterator אחד עבור כל פונקציית generator.<br>
    כל אחד מה־generator iterators יחזיק תמונת מצב עצמאית של המקום שבו עצר הסמן ושל ערכי המשתנים.<br>
    ההתנהלות של כל generator iterator תקרה בנפרד, ולא תושפע בשום צורה מה־generator iterators האחרים.<br>
</p>

### <span style="text-align: right; direction: rtl; float: right; clear: both;">הבדלי מינוח</span>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    בשלב זה יש לנו הרבה מבנים שאפשר לרוץ עליהם, וכל המינוח סביב עניין האיטרביליות נעשה מעט מבלבל.<br>
    ננסה לעשות סדר בדברים:
</p>

<dl style="text-align: right; direction: rtl; float: right; clear: both;">
<dt>Iterable</dt><dd>
    אם ערך מסוים הוא iterable, אפשר לפרק אותו ליחידות קטנות יותר, ולהתייחס לכל יחידה בנפרד.
</dd>
<dt>Iteration, חִזְרוּר</dt><dd>ביצוע יחיד של גוף הלולאה עבור ערך מסוים.</dd>
<dt>Iterator</dt><dd>
    ערך שמייצג זרם של מידע, ומתוכו מאחזרים ערכים אחרים. אפשר לאחזר ממנו ערך אחד בכל פעם, לפי סדר מסוים, בעזרת <code dir="ltr">next()</code>.<br>
    iterator הוא בהכרח iterable, אך לא כל iterable הוא iterator.
</dd>
<dt>Sequence</dt>
    <dd>
        כל iterable שאפשר לחלץ ממנו איברים באמצעות פנייה למיקום שלהם (<code>iterable[0]</code>), כמו מחרוזות, רשימות ו־tuple־ים.<br>
        sequence הוא בהכרח iterable, אך לא כל iterable הוא sequence.
    </dd>
<dt>פונקציית ה־generator</dt><dd>פונקציה המכילה <code>yield</code> ומגדירה אילו ערכים יוחזרו מה־generator.</dd>
<dt>Generator iterator</dt>
    <dd>
        iterator שנוצר מתוך פונקציית ה־generator.
    </dd>
<dt>Generator</dt><dd>לרוב מתייחס לפונקציית ה־generator, אך יש פעמים שמשתמשים במינוח כדי להתייחס ל־generator iterator.</dd>
</dl>

## <span style="text-align: right; direction: rtl; float: right; clear: both;">סיכום</span>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    Generators הם פונקציות שמאפשרות לנו להחזיר סדרות ערכים באופן מדורג.<br>
    כשנקרא לפונקציית generator, היא תחזיר לנו generator iterator שישמש מעין "סמן".<br>
    הסמן ישמור "מצב" שמתאר את המקום שבו אנחנו שוהים בתוך הפונקציה, ואת הערכים שחושבו במהלך ריצתה עד כה.<br>
    בכל שלב, נוכל לבקש את הערך הבא ב־generator בעזרת קריאה לפונקציה <code>next</code> על ה־generator iterator.<br>
    נוכל גם להשתמש במבנים שיזיזו עבורנו את הסמן, כמו <code>for</code> או המרה לסוג נתונים אחר שגם הוא iterable.
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    ל־generators יתרונות רבים:
</p>

<ul style="text-align: right; direction: rtl; float: right; clear: both;">
    <li>אפשר ליצור בעזרתם פונקציות שמחזירות מספר אין־סופי של נתונים.</li>
    <li>במקרים מסוימים, נוכל להיעזר בהם כדי לקבל רק חלק מהתוצאות בכל זמן נתון.</li>
    <li>שימוש נכון בהם יכול להיות מפתח לחיסכון משמעותי במשאבי התוכנית.</li>
</ul>

## <span style="text-align: right; direction: rtl; float: right; clear: both;">תרגילים</span>

### <span style="text-align: right; direction: rtl; float: right; clear: both;">פיצוץ אוכלוסין</span>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    קראו בוויקיפדיה על דרך החישוב של <a href="https://he.wikipedia.org/wiki/%D7%A1%D7%A4%D7%A8%D7%AA_%D7%91%D7%99%D7%A7%D7%95%D7%A8%D7%AA#%D7%A1%D7%A4%D7%A8%D7%AA_%D7%91%D7%99%D7%A7%D7%95%D7%A8%D7%AA_%D7%91%D7%9E%D7%A1%D7%A4%D7%A8_%D7%94%D7%96%D7%94%D7%95%D7%AA_%D7%91%D7%99%D7%A9%D7%A8%D7%90%D7%9C">ספרת הביקורת</a> במספרי הזהות בישראל.
</p>

<ol style="text-align: right; direction: rtl; float: right; clear: both;">
    <li>ממשו פונקציה שמקבלת מספר זהות ללא ספרת ביקורת, ומחזירה את ספרת הביקורת.</li>
    <li>ממשו תוכנית המדפיסה את כל מספרי הזהות האפשריים במדינת ישראל. השתמשו בקוד שכתבתם בסעיף הראשון.</li>
</ol>

### <span style="text-align: right; direction: rtl; float: right; clear: both;">מנה מושלמת לחלוקה</span>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    אפשר לחלק רול סושי של 6 יחידות לאדם אחד, ל־2 אנשים, ל־3 אנשים ול־6 אנשים.<br>
    נתעלם ממצבים שבהם כל אדם מקבל רק חתיכת סושי אחת. זה נשמע לי עצוב.<br>
    נגדיר "מנה מושלמת לחלוקה" כמנה שאם נסכום את כל הצורות לחלק אותה, נקבל את גודל המנה עצמה.
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    לדוגמה:
</p>

<ul style="text-align: right; direction: rtl; float: right; clear: both;">
    <li>רול סושי בעל 6 יחידות הוא מנה מושלמת לחלוקה, כיוון שאפשר לחלק אותו לאדם 1, ל־2 אנשים או ל־3 אנשים. $1+2+3=6$.</li>
    <li>רול סושי בעל 8 יחידות הוא לא מנה מושלמת לחלוקה, כי אפשר לחלק אותו לאדם 1, ל־2 אנשים או ל־4 אנשים. $1+2+4 \neq 8$.</li>
    <li>רול בעל 12 יחידות גם הוא לא מנה מושלמת לחלוקה – $1 + 2 + 3 + 4 + 6 \neq 12$.</li>
    <li>רול בעל 28 יחידות הוא בהחלט מנה מושלמת לחלוקה – $1 + 2 + 4 + 7 + 14 = 28$.</li>
</ul>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    כתבו תוכנית שמדפיסה באופן אין־סופי את כל גודלי המנות שנחשבים מושלמים לחלוקה.
</p>

In [61]:
def perfect_sharing_portion_generator():
  curr_portion = 2

  while True:
    i = curr_portion
    sum = 0

    while i > 1:
      if (curr_portion / i).is_integer():
        sum += curr_portion / i
      i -= 1

    if sum == curr_portion:
      yield curr_portion

    curr_portion += 1




In [62]:
g = perfect_sharing_portion_generator()
for _ in range(10):
  print(next(g))

6
28
496
8128


KeyboardInterrupt: 

### <span style="text-align: right; direction: rtl; float: right; clear: both;">לחששנית</span>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    בקובץ resources/logo.jpg מופיע לוגו הקורס, ובתוכו מוכמנים מסרים סודיים אחדים.<br>
    המסרים הם מחרוזות באורך 5 אותיות לפחות, כתובים באותיות אנגליות קטנות בלבד ומסתיימים בסימן קריאה.
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    פתחו את הלוגו לקריאה בתצורה בינארית, וחלצו ממנו את המסרים הסודיים.<br>
    זכרו שהקובץ עלול להיות גדול מאוד, ועדיף שלא לקרוא את כולו במכה אחת.<br>
    מצאו באינטרנט עזרה בנוגע לפתיחת קבצים בצורה בינארית ולקריאה מדורגת של הקובץ.<br>
    הקפידו שלא להשתמש בטכניקות שלא למדנו (או להוסיף אותן רק בנוסף לפתרון שכזה).
</p>